### PR TITLE
Ryna - L2072.ag_water_irr_mgmt

### PR DESCRIPTION
**Pull request by rynacui**
#504

### CODE REVIEW

File: R/zchunk_L2072.ag_water_irr_mgmt.R [hunk 1]
```diff
@@ -1,67 +1,333 @@
...
+    # For the model, low profit rates are fine as long as it's not the dominant crop in a nest where bioenergy is allowed
+    # Define minimum profit
+    minProfit <- min(L2072.AgCoef_IrrWaterWdraw_ag_mgmt["calPrice"] - L2072.AgCoef_IrrWaterWdraw_ag_mgmt["nonLandVariableCost"]) / 2
+    L2072.AgCoef_IrrWaterWdraw_ag_mgmt %>%
+      # Adjust profit floor
+      mutate(coefficient = round(pmin(coefficient, (calPrice - nonLandVariableCost - minProfit) / WaterPrice), aglu.DIGITS_CALOUTPUT)) %>%
+      # Remove the unnecessary columns
+      select(one_of(LEVEL2_DATA_NAMES[["AgCoef"]])) ->
+      L2072.AgCoef_IrrWaterWdraw_ag_mgmt
+    # # Print out the number of values being changed in each year
```
Comment (rynacui) [inline R/zchunk_L2072.ag_water_irr_mgmt.R]: I commented out these lines for checking changed values for now. How do we want to keep this or not?

File: R/constants.R [hunk 2]
```diff
@@ -116,6 +116,10 @@ aglu.BIO_GRASS_YIELD_KGCM2 <- 0.34
 aglu.BIO_TREE_FERT_IO_GNM2 <- 3.36
 aglu.BIO_TREE_YIELD_KGCM2 <- 0.345
 
+# Water characteristics for biomass (from Vaibhav's paper)
```
Comment (bpbond) [inline R/constants.R]: What paper is this? Can we get a more complete reference?

File: R/constants.R [hunk 3]
```diff
@@ -267,7 +271,7 @@ MAPPED_WATER_TYPES_SHORT            <- c("C", "W")
 names(MAPPED_WATER_TYPES_SHORT)     <- MAPPED_WATER_TYPES
 DEFAULT_UNLIMITED_WATER_PRICE       <- 0
 DEFAULT_UNLIMITED_WITHD_WATER_PRICE <- 0.001
-
+DEFAULT_UNLIMITED_IRR_WATER_PRICE <- 0.001
```
Comment (bpbond) [inline R/constants.R]: Units?

File: R/zchunk_L2072.ag_water_irr_mgmt.R [hunk 4]
```diff
@@ -1,67 +1,333 @@
 #' module_aglu_L2072.ag_water_irr_mgmt
 #'
-#' Briefly describe what this chunk does.
+#' Calculate agriculture water IO coefficients by region / crop / year / GLU / technology.
```
Comment (bpbond) [inline R/zchunk_L2072.ag_water_irr_mgmt.R]: What does IO stand for here?

File: R/zchunk_L2072.ag_water_irr_mgmt.R [hunk 5]
```diff
@@ -1,67 +1,333 @@
...
-    #
-    #
-    # NOTE: This code uses repeat_and_add_vector
-    # This function can be removed; see https://github.com/JGCRI/gcamdata/wiki/Name-That-Function
-    # ===================================================
+    GCAM_region_names <- get_data(all_data, "common/GCAM_region_names")
+    basin_to_country_mapping <- get_data(all_data, "water/basin_to_country_mapping")
+    L132.ag_an_For_Prices <- get_data(all_data, "L132.ag_an_For_Prices")
+    L161.ag_irrProd_Mt_R_C_Y_GLU <- get_data(all_data, "temp-data-inject/L161.ag_irrProd_Mt_R_C_Y_GLU") %>%
+      gather(year, value, -GCAM_region_ID, -GCAM_commodity, -GLU) %>%  # temporary
```
Comment (bpbond) [inline R/zchunk_L2072.ag_water_irr_mgmt.R]: Doesn't matter in this case, but `gather(year, value, matches(YEAR_PATTERN))` would be clearer.

File: R/zchunk_L2072.ag_water_irr_mgmt.R [hunk 6]
```diff
@@ -1,67 +1,333 @@
...
+    L165.GreenRfd_m3kg_R_C_GLU <- get_data(all_data, "L165.GreenRfd_m3kg_R_C_GLU")
+    L165.ag_IrrEff_R <- get_data(all_data, "L165.ag_IrrEff_R")
+    L2052.AgCost_ag_irr_mgmt <- get_data(all_data, "L2052.AgCost_ag_irr_mgmt")
+    L2052.AgCost_bio_irr_mgmt <- get_data(all_data, "L2052.AgCost_bio_irr_mgmt")
+    A03.sector <- get_data(all_data, "water/A03.sector")
+
+    # Primary crops
+    # Compute irrigation water consumption IO coefficients (km3/Mt crop = m3/kg) by region / irrigated crop / year / GLU / management level
+    L165.BlueIrr_m3kg_R_C_GLU %>%
+      ungroup() %>%
```
Comment (bpbond) [inline R/zchunk_L2072.ag_water_irr_mgmt.R]: Is this coming in grouped? if so let's fix it upstream in the chunk that's producing this object.

File: R/zchunk_L2072.ag_water_irr_mgmt.R [hunk 7]
```diff
@@ -1,67 +1,333 @@
...
+      filter(coefficient > 0) %>%
+      # Assume coefs stay constant, copy to all model years
+      repeat_add_columns(tibble::tibble(year = MODEL_YEARS)) %>%
+      # Standardize irrigation water input names
+      mutate(water_sector = "Irrigation",
+             minicam.energy.input = set_water_input_name(water_sector, water_type, A03.sector, GLU = GLU_name)) %>%
+      select(one_of(LEVEL2_DATA_NAMES[["AgCoef"]])) ->
+      L2072.AgCoef_IrrBphysWater_ag_mgmt
+
+    # Compute biophysical water consumption IO coefficients (km3/Mt crop = m3/kg) by region / rainfed crop / year / GLU / management level
```
Comment (bpbond) [inline R/zchunk_L2072.ag_water_irr_mgmt.R]: Seems like the exact same logic is happening here three times in a row? If so let's create an internal function to handle this.

File: R/zchunk_L2072.ag_water_irr_mgmt.R (hunk 1 above)
Comment (bpbond) [inline R/zchunk_L2072.ag_water_irr_mgmt.R]: No, thanks.

**Comment by bpbond** [comment]
Also if any of the comparison of temp-data-inject files you added is bigger than 1 MB please compress.

File: R/constants.R (hunk 2 above)
Comment (rynacui) [inline R/constants.R]: Tried finding the paper but not sure what it is exactly. Could @pkyle or @pralitp help?

File: R/zchunk_L2072.ag_water_irr_mgmt.R [hunk 8]
```diff
@@ -1,67 +1,322 @@
...
+      filter(coefficient > 0) %>%
+      # Assume coefs stay constant, copy to all model years
+      repeat_add_columns(tibble::tibble(year = MODEL_YEARS)) ->
+      L2072.AgCoef_Water_ag_mgmt
+
+    # Separate irrigation water consumption IO coefficients
+    L2072.AgCoef_Water_ag_mgmt %>%
+      filter(water_type == "water consumption") %>%
+      # Standardize irrigation water input names
+      mutate(water_sector = "Irrigation",
```
Comment (rynacui) [inline R/zchunk_L2072.ag_water_irr_mgmt.R]: This step takes ten times longer when I tried it on the combined data above, so I did it separately instead.

File: R/constants.R (hunk 2 above)
Comment (pkyle) [inline R/constants.R]: DOI 10.1007/s11027-013-9497-4
Chaturvedi et al. 2015, Climate mitigation policy implications for global irrigation water demand, Mitig Adapt Strateg Glob Change (2015) 20:389-407.

File: R/zchunk_L2072.ag_water_irr_mgmt.R (hunk 8 above)
Comment (bpbond) [inline R/zchunk_L2072.ag_water_irr_mgmt.R]: Hi @rynacui - can you clarify? Why something take 10x longer? I just want to understand what you've tried, and what you haven't. Thanks.

Comment (rynacui) [inline R/zchunk_L2072.ag_water_irr_mgmt.R]: I mean the function `set_water_input_name`. Although it is repeated for the three data sets (`L2072.AgCoef_IrrWaterCons_ag_mgmt`, `L2072.AgCoef_IrrBphysWater_ag_mgmt`, and `L2072.AgCoef_RfdBphysWater_ag_mgmt`), when I tried it on the combined data `L2072.AgCoef_Water_ag_mgmt`, it takes like 10 times longer to process.

Comment (bpbond) [inline R/zchunk_L2072.ag_water_irr_mgmt.R]: Huh, interesting. I wonder what's going on with that. Thanks.

File: R/zchunk_L2072.ag_water_irr_mgmt.R [hunk 9]
```diff
@@ -1,67 +1,322 @@
...
+      select(one_of(LEVEL2_DATA_NAMES[["AgCoef"]])) ->
+      L2072.AgCoef_IrrWaterCons_bio_mgmt
+
+    # Create tables for water withdrawals IO coefs (consumption divided by irrigation efficiency)
+    L165.ag_IrrEff_R %>%
+      left_join_error_no_match(GCAM_region_names, by = "GCAM_region_ID") %>%
+      right_join(bind_rows(L2072.AgCoef_IrrWaterCons_ag_mgmt, L2072.AgCoef_IrrWaterCons_bio_mgmt), by = "region") %>%
+      # Calculate water withdrawals using consumption divided by irrigation efficiency
+      mutate(coefficient = round(coefficient / field.eff, aglu.DIGITS_CALOUTPUT),
+             # Replace the last character "C" with "W" for input name
```
Comment (kvcalvin) [inline R/zchunk_L2072.ag_water_irr_mgmt.R]: ? Where does the C come from?

File: R/zchunk_L2072.ag_water_irr_mgmt.R [hunk 10]
```diff
@@ -1,67 +1,322 @@
...
+      # Remove extra columns
+      select(one_of(LEVEL2_DATA_NAMES[["AgCoef"]])) ->
+      L2072.AgCoef_IrrWaterWdraw_mgmt
+
+    # Separate table of water withdrawals IO coefs for dedicated bioenergy crops
+    L2072.AgCoef_IrrWaterWdraw_mgmt %>%
+      filter(AgSupplySector == "biomass") ->
+      L2072.AgCoef_IrrWaterWdraw_bio_mgmt
+
+    # Ad hoc adjustment to water coefficients so that none of the region/GLU/crops have negative profit
```
Comment (kvcalvin) [inline R/zchunk_L2072.ag_water_irr_mgmt.R]: Can we open an issue to think this through further? It is definitely not for V1 (and possibly not for v2). But, I'd like to revisit this at some point and don't want to forget.

File: R/zchunk_L2072.ag_water_irr_mgmt.R [hunk 11]
```diff
@@ -1,67 +1,322 @@
...
+      # Calculate water price and water cost
+      mutate(WaterPrice = DEFAULT_UNLIMITED_IRR_WATER_PRICE / conveyance.eff,
+             WaterCost = coefficient * WaterPrice) %>%
+      # Match in non-land variable costs
+      left_join_error_no_match(L2052.AgCost_ag_irr_mgmt, by = c("region", "AgSupplySector", "AgSupplySubsector", "AgProductionTechnology", "year")) %>%
+      # Match in commodity price
+      left_join_error_no_match(L132.ag_an_For_Prices, by = c("AgSupplySector" = "GCAM_commodity")) %>%
+      # Calculate profit as commodity price minus water costs and non-land variable costs
+      mutate(Profit = calPrice - WaterCost - nonLandVariableCost) ->
+      L2072.AgCoef_IrrWaterWdraw_ag_mgmt
```
Comment (kvcalvin) [inline R/zchunk_L2072.ag_water_irr_mgmt.R]: This seems like the wrong name for this tibble. Wouldn't `Profit` be better than `Coef`?

Comment (kvcalvin) [inline R/zchunk_L2072.ag_water_irr_mgmt.R]: Also, do we use the `Profit` column?

File: R/zchunk_L2072.ag_water_irr_mgmt.R [hunk 12]
```diff
@@ -1,67 +1,322 @@
...
+                     "L165.BlueIrr_m3kg_R_C_GLU",
+                     "temp-data-inject/L161.ag_irrProd_Mt_R_C_Y_GLU",
+                     "water/A03.sector") ->
+      L2072.AgCoef_IrrWaterCons_ag_mgmt
+
+    L2072.AgCoef_IrrWaterWdraw_ag_mgmt %>%
+      add_title("Irrigation water withdrawals IO coefficients by region / irrigated crop / year / GLU / management level") %>%
+      add_units("km3/Mt") %>%
+      add_comments("Withdrawals coefs are calculated as consumption coefs divided by irrigation efficiency") %>%
+      add_comments("The same IO coefficients are assigned to both high and low management and keep constant for all model years") %>%
```
Comment (kvcalvin) [inline R/zchunk_L2072.ag_water_irr_mgmt.R]: Maybe mention that we set a floor on profit and adjust the coefficient to ensure that it is met?

File: R/zchunk_L2072.ag_water_irr_mgmt.R (hunk 9 above)
Comment (rynacui) [inline R/zchunk_L2072.ag_water_irr_mgmt.R]: line 114 `set_water_input_name`, for water consumption

File: R/zchunk_L2072.ag_water_irr_mgmt.R (hunk 11 above)
Comment (rynacui) [inline R/zchunk_L2072.ag_water_irr_mgmt.R]: You're right, it is not used.

Comment (kvcalvin) [inline R/zchunk_L2072.ag_water_irr_mgmt.R]: Okay, given that, disregard my comment about re-naming the tibble.

**Pull request merged**